### PR TITLE
[Feature:InstructorUI] Autograding Status UI

### LINKED ISSUE #6765: Grading Done Web GUI
**What problem are you trying to solve with Submitty**
Currently, instructors can monitor the autograding queue from the ssh command line using the 
grading_done.py script.  The script prints stats once every 5 seconds.  As you leave the script running over 
time, you can estimate the progress and pace.

Instructors must have limited sudo access in order to run this script.
Users who do not have ssh access, or who are not comfortable using using the command line
are not able to check the grading queue progress details.

**Describe the way you'd like to solve this problem**
Let's make a webpage version of this data.
Same information, and also tracking the progress and pace if the page is left open for a period of time.
[NOTE - we will need to be careful, this page might be expensive & a potential large load on the file system since the queue is currently stored as files.  We do eventually want to move the queue to the database, but we aren't there yet.]
It will have some overlap with the work-in-progress docker info page
(capabilities & worker machines)

### PR DESCRIPTION
**Pull request by HenrikLam**
### What is the current behavior?
Closes #6765

### What is the new behavior?
New page that allows user to monitor the status of autograding, each tables are updated continuously.

### Other information?

You can test by hitting the regrade button on each gradeable, or by making submissions on a gradeable. There also seems to be gradeable being batch submitted when you do vagrant up. You can also do `sudo systemctl stop submitty_autograding_shipper`
`sudo systemctl stop submitty_autograding_worker` to stop the system from finishing grading the files to inspect the UI closely.

### CODE REVIEW

File: site/app/controllers/AutogradingStatusController.php
Comment (cjreed121) [inline site/app/controllers/AutogradingStatusController.php:20]: You could give this a return type of ResponseInterface.

File: site/app/controllers/AutogradingStatusController.php
Comment (cjreed121) [inline site/app/controllers/AutogradingStatusController.php:23]: Potentially a quicker check could be to call getInstructorLevelAccessCourse and see if there are any results rather than collecting all instructors and faculty.

File: site/app/controllers/AutogradingStatusController.php
Comment (cjreed121) [inline site/app/controllers/AutogradingStatusController.php:39]: Put a / in front of route as all others have that

File: site/app/libraries/QueueItem.php
Comment (cjreed121) [inline site/app/libraries/QueueItem.php:10]: `@prop-read` does not do anything here as this is not a model. This probably could just be a model rather than a library since it just stores data for you.

File: site/app/views/AutogradingStatusView.php
Comment (cjreed121) [inline site/app/views/AutogradingStatusView.php:17]: This should get a breadcrumb because top left just says Submitty

File: site/public/js/autograding-status.js
Comment (cjreed121) [inline site/public/js/autograding-status.js:15]: Rather than telling eslint to ignore these at the top you can put:

/* global buildUrl, csrfToken */

That will tell eslint that it is not in current scope but is global when the browser executes the code.

Also csrfToken is not necessary here as this is a get request. Csrf tokens are only required on post requests.

File: site/public/js/autograding-status.js
Comment (cjreed121) [inline site/public/js/autograding-status.js:22]: You can simply just add this to the global list as mentioned above.

**Review by cjreed121 — CHANGES_REQUESTED** [review]

File: site/app/libraries/GradingQueue.php
Comment (williamjallen) [inline site/app/libraries/GradingQueue.php:421]: Do we want the entire stack trace here?

File: site/app/libraries/GradingQueue.php
Comment (williamjallen) [inline site/app/libraries/GradingQueue.php:340]: Is this temporary variable necessary if we're only using it in one spot?

File: site/app/libraries/GradingQueue.php
Comment (williamjallen) [inline site/app/libraries/GradingQueue.php:356]: Can we be certain that there will always be a capability named "default"?  (I'm not too familiar with how this works yet)

File: site/app/libraries/GradingQueue.php
Comment (williamjallen) [inline site/app/libraries/GradingQueue.php:389]: `!==` might be better here

File: site/app/libraries/GradingQueue.php
Comment (williamjallen) [inline site/app/libraries/GradingQueue.php:392]: `===` might be better here

File: site/app/libraries/QueueItem.php
Comment (williamjallen) [inline site/app/libraries/QueueItem.php:1]: Should a few basic unit tests be added for this file?

File: site/app/templates/AutogradingStatus.twig
Comment (williamjallen) [inline site/app/templates/AutogradingStatus.twig:102]: The bootstrap class `.text-left` would do the same thing and remove the need for an inline style.  Same goes for a number of other places in this file.

File: site/app/templates/AutogradingStatus.twig
Comment (williamjallen) [inline site/app/templates/AutogradingStatus.twig:93]: Do we need a custom border class here (and elsewhere in this table) if we have the `.table-bordered` bootstrap class?

File: site/app/templates/AutogradingStatus.twig
Comment (williamjallen) [inline site/app/templates/AutogradingStatus.twig:174]: It might be cleaner if this section was indented a little bit.

File: site/public/css/autograding-status.css
Comment (williamjallen) [inline site/public/css/autograding-status.css:18]: What purpose are these trying to serve?  It seems to me like we're writing HTML in a CSS file here...

File: site/public/css/server.css
Comment (williamjallen) [inline site/public/css/server.css:2033]: Have you tested these to make sure they work on all browsers?  Is there any reason why these are necessary instead of just relying on native OS scrollbars?  Scrollbars are tricky because they look, and act, differently on different systems and browsers.

File: site/public/js/autograding-status.js
Comment (williamjallen) [inline site/public/js/autograding-status.js:29]: `!==` here perhaps, as well as on line 32?

File: site/public/js/autograding-status.js
Comment (williamjallen) [inline site/public/js/autograding-status.js:84]: `===`?

File: site/public/js/autograding-status.js
Comment (williamjallen) [inline site/public/js/autograding-status.js:103]: Instead of using setInterval, it might be better to use setTimeout() and then call that recursively, waiting for requests to come back before sending out the next one.  That way, if requests are taking a long time, we don't spam the server with requests.

File: site/public/js/autograding-status.js
Comment (williamjallen) [inline site/public/js/autograding-status.js:129]: Same comment as above regarding setInterval()

**Review by williamjallen — CHANGES_REQUESTED** [review]
Code looks good in general, although I left a few comments throughout.  I would ideally like to see a few unit tests for this.  It's a new page and it would be good to ensure that it stays functional instead of attempting to retrofit tests to the code later.  Perhaps a couple e2e tests to check the page update toggle, as well as basic checking to make sure access controls work as expected for each type of user.

File: site/app/controllers/AutogradingStatusController.php
Comment (HenrikLam) [reply site/app/controllers/AutogradingStatusController.php:15]: As far as I know,`@AccessControl` only checks your access for the current course. However, this page doesn't exist as a course page so `AccessControl` would not work.

Comment (williamjallen) [reply site/app/controllers/AutogradingStatusController.php:15]: Hmm, ok.  I'm still a little wary of this because someone could easily come in later and make changes, not knowing that access control had to be done in each function, and unknowingly leave a security vulnerability open.  It's tough though so if there's no better option I'll approve it the way it is.

File: site/app/libraries/GradingQueue.php
Comment (HenrikLam) [reply site/app/libraries/GradingQueue.php:421]: An entire stack trace might be a bit excessive, I assume you are referring to the appending of `$e`?

File: site/app/libraries/GradingQueue.php
Comment (HenrikLam) [reply site/app/libraries/GradingQueue.php:340]: Which variable are you referring to? Both `$entry` and `$regrade` are used later in the code.

File: site/app/libraries/GradingQueue.php
Comment (HenrikLam) [reply site/app/libraries/GradingQueue.php:356]: Default should exist from the submitty setup script, it is possible that someone get rid of it if they go into the system itself however.

File: site/app/templates/AutogradingStatus.twig
Comment (HenrikLam) [reply site/app/templates/AutogradingStatus.twig:93]: The border is for the borders between the columns, which does not exist between every column. For example, between the machines and capabilities, each capability is a column but for the sake of association, I did not put a border between them.
image

File: site/app/libraries/GradingQueue.php
Comment (williamjallen) [reply site/app/libraries/GradingQueue.php:421]: Yeah, I was unclear.  Currently it will print a large stack trace.  What I meant to suggest is that it be switched to `$e->getMessage()` to only print the message and not the entire stack trace.

File: site/app/templates/AutogradingStatus.twig
Comment (HenrikLam) [reply site/app/templates/AutogradingStatus.twig:174]: Good point, I never know how twig should look like styling wise. I have seen both indenting and not indenting done but I was never sure.

File: site/public/css/autograding-status.css
Comment (HenrikLam) [reply site/public/css/autograding-status.css:18]: This allows the table to be displayed properly when the windows become too small. Without these lines, the table displayed will consist of `f21`, `sample`, `grading_homework`, and so on without any of the column names,
image

File: site/public/css/server.css
Comment (HenrikLam) [reply site/public/css/server.css:2033]: It seems to work for both Chrome and Firefox, I am only using it because I hate the way the native windows scrollbar looks.

File: site/app/libraries/GradingQueue.php
Comment (williamjallen) [reply site/app/libraries/GradingQueue.php:340]: I was referring to `$regrade` here but I see that it's used in more places than I thought initially so it looks ok.

File: site/public/js/autograding-status.js
Comment (HenrikLam) [reply site/public/js/autograding-status.js:103]: I considered using `setTimeout()`, but then the button for pausing the timeout becomes tricky, or at least I don't know of a way to get rid of things queued to run with `setTimeout()`. With setInterval, however, I can use the `clearInterval()`.

Comment (williamjallen) [reply site/public/js/autograding-status.js:101]: An easy way to do this would be to use the button to add/remove an attribute somewhere and then check for the existence of that attribute before calling `setTimeout()` again.

File: site/public/css/autograding-status.css
Comment (williamjallen) [reply site/public/css/autograding-status.css:18]: Oh, ok, sounds good then.  Might want to leave a comment near the table saying that this has been done, just in case someone comes along later and wants to edit this table.

File: site/app/controllers/AutogradingStatusController.php
Comment (HenrikLam) [reply site/app/controllers/AutogradingStatusController.php:16]: Yeah I agree it isn't the greatest solution. I should also note that something similar is also done in the docker UI page, not to justify this but as a note of where else this is happening.

File: site/public/js/autograding-status.js
Comment (HenrikLam) [reply site/public/js/autograding-status.js:101]: What are the advantages of using `setTimeout()` over `setInterval()`, right now `setInterval()` also allows us to wait for the response to come back before sending another request.

File: site/app/libraries/QueueItem.php
Comment (HenrikLam) [reply site/app/libraries/QueueItem.php]: Unit tests might be a little tricky since it is reading the queue files which are constantly being grabbed by the `autograding_workers`, it could be done if we first manually stop the shippers and workers though.

Comment (cjreed121) [reply site/app/libraries/QueueItem.php]: Unit tests should not rely on autograding or the website to be functional. Unit tests are run before our e2e tests. Instead you can create very basic json files and call the constructor with that.

File: site/public/js/autograding-status.js
Comment (williamjallen) [reply site/public/js/autograding-status.js:101]: It would just mean that you wouldn't have to always be clearing/resetting the interval.  If I understand the code correctly, this code essentially uses `setInterval()` to do the job of `setTimeout()`.  `setTimeout()` only runs once after a set amount of time, while the current code sets the interval and runs the function once which clears the interval and then re-sets it.

File: site/app/controllers/AutogradingStatusController.php
Comment (cjreed121) [inline site/app/controllers/AutogradingStatusController.php:44]: This can also be shortened like mentioned as above.

File: site/app/controllers/GlobalController.php
Comment (cjreed121) [inline site/app/controllers/GlobalController.php:393]: Here as well

File: site/app/libraries/GradingQueue.php
Comment (cjreed121) [inline site/app/libraries/GradingQueue.php:328]: Maybe make this function header multiple lines? Also where it gets called too seems long so maybe do the same there too.

File: site/app/libraries/database/DatabaseQueries.php
Comment (cjreed121) [inline site/app/libraries/database/DatabaseQueries.php:3855]: Why select distinct? This table has a defined primary key on semester, course, and user_id so there is no way for there to be multiple identical entries per user.

File: site/app/views/AutogradingStatusView.php
Comment (cjreed121) [inline site/app/views/AutogradingStatusView.php:10]: You could put a type hint on $progress.

File: site/public/js/autograding-status.js
Comment (cjreed121) [inline site/public/js/autograding-status.js:13]: The csrf token is not necessary as this is a GET request

**Review by cjreed121 — CHANGES_REQUESTED** [review]

File: site/app/libraries/GradingQueue.php
Comment (HenrikLam) [reply site/app/libraries/GradingQueue.php:328]: Is there a way to make the linter happy without using a line for each parameter? I opted for this way unoriginally because I thought it will become too big with one line each for each parameter.

Comment (williamjallen) [reply site/app/libraries/GradingQueue.php:328]: I don't know of any way to satisfy the linter without either putting them all on one line or each on its own line.

File: site/public/css/server.css
Comment (williamjallen) [reply site/public/css/server.css:2033]: Ok, I'm still a little wary of it but it does seem to work as expected on safari and chrome.  It does look a little weird in light mode though.  You might want to have the background color update with the theme.

**Review by williamjallen — COMMENTED** [review]
Most major changes have been addressed beyond the comments Chris left so I think it's close to being ready to merge once those are addressed.  I do still have a couple concerns about relying upon the existence of the "default" capability but perhaps we can discuss that in the meeting tomorrow.  I would also really like to see a little bit of testing where possible to check for things like making sure that students can't access this page for example, as well as some unit testing.  It doesn't necessarily have to be done in this PR but I just want to make sure that we aren't adding yet another page which ultimately doesn't get tests run against it on a regular basis.

Also, the pause updates button doesn't seem to work for me.  Could it be that something was forgotten somewhere when `setInterval()` was changed to `setTimeout()`?

File: site/app/libraries/GradingQueue.php
Comment (cjreed121) [reply site/app/libraries/GradingQueue.php:328]: I'm pretty sure you would have to put all arguments on separate line. I think that would still look much better than having 1 very long line.

File: site/app/controllers/GlobalController.php
Comment (cjreed121) [inline site/app/controllers/GlobalController.php]: No reason to run this query again if you can just store it in a variable from your query above on 394.

**Review by cjreed121 — CHANGES_REQUESTED** [review]

**Comment by HenrikLam** [comment]
I left out the flag to stop the update thinking stopTimeout would be enough, but it seems sometimes this breaks when we are unlucky with the timing. It seems to work consistently for me now after adding the flag.

**Review by cjreed121 — CHANGES_REQUESTED** [review]

File: site/app/controllers/AutogradingStatusController.php
Comment (cjreed121) [reply site/app/controllers/AutogradingStatusController.php:16]: Maybe just add a comment saying this is a security check. I have seen other places in the code do that when a manual security check was needed.

File: site/app/controllers/AutogradingStatusController.php
Comment (cjreed121) [inline site/app/controllers/AutogradingStatusController.php:54]: Is there a better name for this besides $tmp?

**Review by cjreed121 — CHANGES_REQUESTED** [review]
Looks good. About to approve but just one last code style comment.

**Review by cjreed121 — APPROVED** [review]

**Review by williamjallen — APPROVED** [review]

**Pull request merged**
